motion_primitives_forward_controller/src/motion_primitives_forward_controller.cpp

```diff
@@ -0,0 +1,554 @@
+            "Primitive %zu invalid: LINEAR_JOINT requires joint_positions.", i);
+          return rclcpp_action::GoalResponse::REJECT;
+        }
+        break;
+
+      case MotionType::LINEAR_CARTESIAN:
+        RCLCPP_INFO(get_node()->get_logger(), "Primitive %zu: LINEAR_CARTESIAN (LIN)", i);
+        if (primitive.poses.empty())
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(),
+            "Primitive %zu invalid: LINEAR_CARTESIAN requires at least one pose.", i);
+          return rclcpp_action::GoalResponse::REJECT;
+        }
+        break;
+
+      case MotionType::CIRCULAR_CARTESIAN:
+        RCLCPP_INFO(get_node()->get_logger(), "Primitive %zu: CIRCULAR_CARTESIAN (CIRC)", i);
+        if (primitive.poses.size() != 2)
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(),
+            "Primitive %zu invalid: CIRCULAR_CARTESIAN requires exactly two poses.", i);
+          return rclcpp_action::GoalResponse::REJECT;
+        }
+        break;
+
+      default:
+        RCLCPP_ERROR(
+          get_node()->get_logger(), "Primitive %zu invalid: unknown motion type %u.", i,
+          primitive.type);
+        return rclcpp_action::GoalResponse::REJECT;
+    }
+  }
+
+  return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+}
+
+rclcpp_action::CancelResponse MotionPrimitivesForwardController::goal_cancelled_callback(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<ExecuteMotion>>)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Canceling goal");
+
+  std::lock_guard<std::mutex> guard(command_mutex_);
+  reset_command_interfaces();
+  robot_stopped_ = true;
+  // send stop command immediately to the hw-interface
+  (void)command_interfaces_[0].set_value(static_cast<double>(MotionType::STOP_MOTION));
+  while (!moprim_queue_.empty())
+  {  // clear the queue
+    moprim_queue_.pop();
+  }
+
+  RCLCPP_INFO(get_node()->get_logger(), "Waiting for Robot to stop...");
+  while (execution_status_ != ExecutionState::STOPPED)
+  {
+    // Small sleep to prevent busy waiting
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+  }
+
+  reset_command_interfaces();
+  // send reset stop command to the hw-interface
+  (void)command_interfaces_[0].set_value(static_cast<double>(MotionType::RESET_STOP));
+  robot_stopped_ = false;
+
+  RCLCPP_INFO(get_node()->get_logger(), "Robot stopped, ready for new motion primitives.");
+  return rclcpp_action::CancelResponse::ACCEPT;
+}
+
+void MotionPrimitivesForwardController::goal_accepted_callback(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<ExecuteMotion>> goal_handle)
+{
+  std::thread{std::bind(&MotionPrimitivesForwardController::execute_goal, this, goal_handle)}
+    .detach();
+}
+
+void MotionPrimitivesForwardController::execute_goal(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<ExecuteMotion>> goal_handle)
+{
+  const auto & goal = goal_handle->get_goal();
+
+  if (robot_stopped_)
```

Comment:
This is missing thread synchronization. It should at least be a `std::atomic<bool>`.